### PR TITLE
Adding AtMegaXXRFR2 support and avr-rss2 platform

### PR DESCRIPTION
**Pull request by herjulf**
* Adding support for Atmel AtMega128RFR2 and AtMega256RFR2
* Adding platform avr-rss2 using AtMega256RFR2

Cooja and avrora patch will come later

Thanks,
Robert, Atis, Voravit, Jens

### CODE REVIEW

**Comment by joachimkr** [comment]
I also use the new AVR for a new project and are very happy, when this device will be fully supported in contiki!
Greetings, Joachim

**Comment by herjulf** [comment]
Yes. As I understood you are using the AtMega256RFR2 with port? Right? The R2 radios have some new features including lower power use.
Thanks.

**Comment by g-oikonomou** [comment]
@joachimkr are you happy with this pull? I've got Atis here with me and he's supportive but I'd appreciate your feedback: more reports are welcome!

**Comment by joachimkr** [comment]
Hi,

I am very happy with it, because it is coming from an experienced author.
Seems, he did a lot with the atmega128rfr1 and really understands that unit.
And I see my first opportunity to select that newer AVR with builtin
advanced radio.
I am in progress using the avr-rss2 branch right now.
During my work, I probably generate more feedback, soon.

Greetings,
Joachim
Am 11.03.2016 21:06 schrieb "George Oikonomou" <notifications@github.com>:

**Comment by herjulf** [comment]
### Comments and background to this work.
Very much what is in the README files.

The Amega128RFA1 has successor AtMegaXXRFR2 which said to be drop-in-replacement for the older version. Of course there are improvements i.e. RPC (reduced power option) different techniques for HW duty cycling and phase shift distance measurements. RX current can be tuned to less than 10mA to give 18mA for the RFA1.  rf230_set_rpc() an new call is added to control RPC. There is small test-example among the platform examples. Of course the new MCU has there own and new signatures. The MCU are needs  to be supported by the toolchain. gcc on ubuntu lacks support. For now you'll need get gcc-tool chain from Atmel's website. There are also Windows support. People have reported success. I'm a linux user and hacker.

The port adds an own register file a cleaned up copy from RFA1 but with some small changes. This is done not to impact with any of the existing platforms. The new register should work all memory variants 
of the chip 64k 128k and 256k flash. The port adds support for  AtMega128RFR2 and  AtMeg256RFR2. Not for AtMeg64RFR2 just because I haven't got access to any chip to test. It should be straight forward 
to add.

To complement and to test MCU support the avr-rss2 platform is added. A  motiivation for the work.  The platform can be used with different MCU's via the MCU= option this  described the the README.md files. The platform includes some examples, among them a IP64/NAT/GW using the enc28j60 module/dongle. Also an sniffer app is included George sensniff or wsnbridge with wireshark. 

The third addition is the travis regression test for the platform actually contributes by Atis who also added recent avr-tooschain for avr-regression tests to support the newer versions.

Avrora is still the blues... We played with support for the new Radio and also tried to support the 256k variant. This quite a work as avorora does not support devices >128k. gcc creates object code has trampoline sections and extened jump call function. For simultion/emulation we've compiled the platform for AtMega128RFR2. The avrora effeors are not ready nor included and needs quite a lot of work.

In summary. This work adds new functionality for new Atmel's MCU's with focus to have clean patches well separated not to interfere with existing platforms nor core functionality.

**Comment by joachimkr** [comment]
The avr gcc from Atmel is working in the cygwin environment. So that barrier is low.

**Comment by g-oikonomou** [comment]
Why do you have examples under the platform dir? Not common practice.

Looks like a very extensive contribution you've made here. Looking forward to trying it!

**Comment by herjulf** [comment]
Well some are more specific for platform i e hello-sensors. follows the of course Contiki frame dev framework but platforms has different sensors. The  avr_radio_power can of course be shared with platform sharing the RFR2 radio but so far only this platform. Feel free to suggest a better place. The rpl-border-router is a Ethernet variant of the slip version that is in Contiki this works adds the SPI-glue needed for AtMega's  WRT enc28j60. Well yes it can be moved or even removed. 

But the goal is to have to a strong separation and safe patch to get things moving forward.

**Comment by g-oikonomou** [comment]
Common practice would be `examples/<this-platform/<that-example>/`. Any reason this can't be used?

**Comment by herjulf** [comment]
> Common practice would be examples/<this-platform/<that-example>/. Any reason this can't be used?
It's absolutely possible....

**Comment by joachimkr** [comment]
I checked: Atmel has **just published the latest** Compilers supporting the newer RFR2 versions for Linux. So my last comment is out of probable blocks, now.
Note: The Atmega128RFRx and Atmega256RFR2 all use the 8-bit compiler.

<Atmel AVR 8-bit Toolchain 3.5.1 - Linux 32-bit
(41.2 MB, updated March 2016)

Atmel AVR 8-bit Toolchain 3.5.1 - Linux 64-bit
(44.9 MB, updated March 2016)

Atmel AVR 32-bit Toolchain 3.4.3 - Linux 64-bit
(26.1 MB, updated September 2015)
Atmel 32-bit Toolchain (3.4.3) 6.2.0.742 - Header Files
(2.51 MB, updated September 2015)>

**Comment by herjulf** [comment]
Discussion stopped, meanwhile code has been used at KTH Royal Inst of Technology Stockholm. Platform including some of the included examples. IP64NAT/GW /Ethernet is used both for MQTT, CoAP, KISS-WSN/sensd and will be further tested and interoped in a GreenIoT workshop on Moday Electrum/Kista  As Atis mentioned radio-code would need a clean up and avrora would need to support the 256k variants of chips. So there are work to be done. Can also mention current preferences for sniffer issenssniff. Was I expected to rework the patch?

**Comment by alignan** [comment]
I would like to have @g-oikonomou comment about moving the examples from `platforms` to the `examples` location, to keep stuff consistent.

**Comment by herjulf** [comment]
Hello,
Moving the examples is no problem. Was I expected to rework the patch? 
FYI. It seems like the AtMegaXXRFR2 chips has timers for decent a TSCH implementation.

**Comment by alignan** [comment]
@herjulf Yes, I think this was the expected.

I have reviewed most of the code and feels OK, looking forward to have the chance and test it on actual hardware :+1: 

(Funny how I'm the one likely merging new platforms xD)

**Comment by herjulf** [comment]
@herjulf Yes, I think this was the expected.

I have reviewed most of the code and feels OK, looking forward to have the chance and test it on actual hardware :+1:



(Funny how I'm the one likely merging new platforms xD)

Hi,
Thanks a lot for reviewing. 
Send me an email and I help with some HW to test.
If it just jmoving the example directory it can be a follow-up patch. 
Can also mention that we're in outdoor RF-test also between platforms.  

BR

**Comment by alignan** [comment]
Thanks for the help! I will write you offline.

I would prefer to have the suggested change already addressed in this PR, so it won't be a pending thing to expect later, I hope it is not an inconvenience.

**Comment by herjulf** [comment]
_I would prefer to have the suggested change already addressed in this PR, so it won't be a pending thing to expect later, I hope it is not an inconvenience._

Ok I'll look into it,,
Thanks

**Comment by g-oikonomou** [comment]
Robert, I see you have moved examples as requested, thanks. The travis test still tries to build from the original location. `regression-tests/23-compile-avr/Makefile` needs updated to reflect the most recent change.

I want to merge this one already, so as soon as you have had a chance to fix this, please notify.

**Comment by herjulf** [comment]
Yes, I missed that.. regression tests are now  updated also some small clean-ups. We have some radio changes in the pipe but those changes will impact all avr-platforms and could need some more testing. So maybe this enough for this PR.  Other comments?

**Comment by g-oikonomou** [comment]
Thanks for addressing this, Robert.

I see there are some places where the code style could be improved. However, this looks like a solid contribution for what appears to be a very good piece of kit. Plus, avr-based platforms have been in a bit of a state for a while now and lacking active maintenance. It's good to get some active involvement from  people who understand the h/w.

I' merging this this one already, I'm sure we can start updating code style here and there as we go along.

**Comment by herjulf** [comment]
Thanks a lot for review and support!

**Comment by viktpr** [comment]
This is great, thank you for this work!

I am using Contiki in my bachelor's thesis and have freely used much of this work to make a new platform based on [ZigBit ATmega256RFR](http://www.atmel.com/devices/ATmega256RFR2-ZigBit-Wireless-Module.aspx) wireless module. 
I only have the preliminary platform support with no radio and network yet, I will publish
it when I get more things working and tested.

**Comment by herjulf** [comment]
Viktor P. writes:
 > This is great, thank you for this work!
 > 
 > I am using Contiki in my bachelor's thesis and have freely used much of this work to make a new platform based on [ZigBit ATmega256RFR](http://www.atmel.com/devices/ATmega256RFR2-ZigBit-Wireless-Module.aspx) wireless module. 
 > I only have the preliminary platform support with no radio and network yet, I will publish
 > it when I get more things working and tested.

 Thanks, Yes this basic support will help a lot of course more work is needed. 
 IMO moodule you're working with seems be straight forward to add. We've some 
 small set of radio clean-up patches in the pipe currently being now tested in 
 Outdoor PDR-testing at Uppsala University meteorological site. 
 TSCH should be possible timers should be sufficient for a good implementation.
 Cooja and avrora for AtMegaXXRFR2 for 64k and 128k chips should be possible
 we have draft code somewhere but needs to get integrated. Also the simulated 
 radio needs some fixes to show correct power use this lower. There are some 
 measurements in the example directory. Atis and I tried to fix avrora for 256k 
 chip but had to give up for other projects. 

 Best regards
					--ro

**Comment by viktpr** [comment]
You are certainly right, I was able to get it running in a few hours with no real issues besides the expected platform configuration differences that are easily resolved by glancing at the documentation and a bit of experimenting. 

It certainly would be nice to have proper AVR platform support in Cooja, something that is greatly missing now. I am looking forward to any further work being done to improve the state
of AVR platforms in Contiki and I hope I can be of help after finishing the schoolwork.

Regards
